### PR TITLE
Filter expression parser implementation.

### PR DESCRIPTION
**Pull request by TristonianJones**
The following change introduces an Antlr4 grammar that defines the filter syntax.
The grammar supports logical operations and sequences of restriction expressions,
with the ability to reference object fields and invoke functions supported by the
API as described in Grafeas Issue https://github.com/grafeas/grafeas/issues/29.

The syntax is equivalent to that supported by Google Cloud Logging [Advanced Filters](https://cloud.google.com/logging/docs/view/advanced-filters)
with one notable exception: it can accurately parse indexing operations on a variable
of the form:

    a[b]   // select field 'b' from identifier 'a'

Lack of support for the above syntax is a known limitation of Advanced Filters which
is not present here.

Parsed filters are output in the [Common Expression Language](https://github.com/google/cel-spec)
abstract syntax representation. A key difference between CEL and list filters is that
list filters may treat barewords in the expression as unquoted text. Within CEL, these
barewords would be flagged as identifiers. Whether the value is a string or identifier is
a contextual decision best left to the filter evaluation step.

Note, the CEL abstract syntax proto and Antlr4 Go runtime are vendored into this PR.

### CODE REVIEW

**Comment by furuholm** [comment]
Great to see progress in this area!

For someone who has not been part of the discussion leading up to this PR: How is this supposed to be integrated with the existing code base? More specifically I'm thinking about the the storage layer implementations. Is there something in place that I missed (there are a lot of new files) or will this be handled separately?

**Comment by R2wenD2** [comment]
@furuholm after we get this PR in, I'll work on a library that uses it to parse filters and map them to fields in the postgresql db.

**Comment by TristonianJones** [comment]
Apologies on the file count @furuholm. The vendored Antlr4 Go runtime and generated source accounts for the bulk of the change. There are probably <1k lines of newly minted code under the filtering/ directory (despite what the contribution stats look like).

Happy to contribute. :)

**Comment by furuholm** [comment]
@R2wenD2 Great!

@TristonianJones No worries! :)

Really looking forward to have great querying capabilities in Grafeas.

File: server-go/filtering/parser/parser_test.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	TestCaseDelimiter    = "\n\n"
+)
+
+type baseline struct {
+	expected *expr.ParsedExpr
+	errors   string
+	source   common.Source
+}
+
+func TestParse_Complex(t *testing.T) {
```
Comment (clrprod) [inline server-go/filtering/parser/parser_test.go]: This set of nested functions seems relatively complicated - is there any way we could just have a single test case which loads all the files and runs the tests inside a series of nested for loops?  (That way you can read the testcase without having to go look at function definitions).

File: server-go/filtering/common/source_test.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	str, found := source.Snippet(1)
+	assertTrue(t, found)
+	assertEquals(t, "hello, world", str)
+	str2, found := source.Snippet(2)
+	assertFalse(t, found)
+	assertEquals(t, "", str2)
+}
+
+// Helper assertions.
+func assertTrue(t *testing.T, actual bool) {
```
Comment (clrprod) [inline server-go/filtering/common/source_test.go]: Can we just inline these functions? I don't think they make the tests more readable. Alternatively can you call t.Helper() so the error messages include the actual line in the test function that causes the error?

Comment (TristonianJones) [reply server-go/filtering/common/source_test.go]: Yes, definitely. I'll inline them.

File: server-go/filtering/parser/parser.go [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+	switch e.(type) {
+	case antlr.InputMisMatchException:
+		errorMsg = "Input mismatch"
+	case antlr.NoViableAltException:
+		errorMsg = "Unexpected token"
+	}
+	p.errors.ReportError(p.source, common.NewLocation(line, column), errorMsg)
+}
+
+func (p *parser) ReportAmbiguity(recognizer antlr.Parser, dfa *antlr.DFA,
```
Comment (R2wenD2) [inline server-go/filtering/parser/parser.go]: Can you add a function level comment here, including a description of why the method is intentionally left empty?

File: server-go/filtering/parser/parser.go [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+	startIndex, stopIndex int, exact bool, ambigAlts *antlr.BitSet,
+	configs antlr.ATNConfigSet) {
+	// Intentional
+}
+
+func (p *parser) ReportAttemptingFullContext(recognizer antlr.Parser,
```
Comment (R2wenD2) [inline server-go/filtering/parser/parser.go]: Same here and below with function level comments.

**Comment by R2wenD2** [comment]
Also, thanks for this, @TristonianJones

File: server-go/filtering/parser/parser_test.go (hunk 1 above)
Comment (clrprod) [reply server-go/filtering/parser/parser_test.go]: If you need a message diffing library https://github.com/d4l3k/messagediff works.

re: IDE's and rerunning tests - I don't care strongly about whether there is a single table of tests, or you load each file in a separate test, I just just strongly encourage having tests be as simple as possible with minimal usage of helper functions and layers of abstraction.

The idea is that you want tests to be 'obviously correct' on inspection. Layers of function calls make it impossible to have this, and instead you have to go read each function and it's easier for bugs to slip in.

However I also approved this, so feel free to ignore my suggestion - this is mostly minor test style and I'm happy the tests exist at all :D

File: server-go/filtering/parser/parser.go (hunk 3 above)
Comment (TristonianJones) [reply server-go/filtering/parser/parser.go]: Sure

File: server-go/filtering/parser/parser.go (hunk 4 above)
Comment (TristonianJones) [reply server-go/filtering/parser/parser.go]: Of course, happy to add more comments.

File: server-go/filtering/parser/parser_test.go (hunk 1 above)
Comment (TristonianJones) [reply server-go/filtering/parser/parser_test.go]: Thanks, I'm not sure I have time to restructure the tests just now, but will keep this in mind for the future.

**Comment by TristonianJones** [comment]
Feel free to merge if you're happy with the responses to feedback.

**Pull request merged**